runtime/src/main/java/org/corfudb/protocols/logprotocol/SMRRecord.java

```diff
@@ -0,0 +1,244 @@
+    @Setter
+    public transient SMRRecordLocator locator = null;
+
+    /**
+     * The global address this record is associated to.
+     */
+    @Getter
+    @Setter
+    long globalAddress = Address.NON_ADDRESS;
+
+    /**
+     * The size in bytes of this SMREntry when serialized.
+     */
+    @Getter
+    private int serializedSize = 0;
+
+    /**
+     * SMRRecord constructor.
+     */
+    public SMRRecord(String smrMethod, @NonNull Object[] smrArguments, ISerializer serializer) {
+        this.SMRMethod = smrMethod;
+        this.SMRArguments = smrArguments;
+        this.serializerType = serializer;
+    }
+
+    /**
+     * If this record is trimmed because of compaction.
+     */
+    public boolean isCompacted() {
+        return this == COMPACTED_RECORD;
+    }
+
+    /**
+     * If this byte array representing the record is
+     * trimmed because of compaction.
+     */
+    public static boolean isCompacted(byte[] record) {
+        ByteBuf buf = Unpooled.wrappedBuffer(record);
+        return buf.readBoolean();
+    }
+
+    /**
+     * Set the upcall result for this entry.
+     */
+    public void setUpcallResult(Object result) {
+        upcallResult = result;
+        haveUpcallResult = true;
+    }
+
+    /**
+     * Set the undo record for this entry.
+     */
+    public void setUndoRecord(Object object) {
+        this.undoRecord = object;
+        undoable = true;
+    }
+
+    /**
+     * Clear the undo record for this entry.
+     */
+    public void clearUndoRecord() {
+        this.undoRecord = null;
+        undoable = false;
+    }
+
+    /**
+     * Deserialize from buffer and return a new {@code SMRRecord}.
+     *
+     * @param b byte buffer to deserialize.
+     * @return an {@code SMRRecord} deserialized from buffer.
+     */
+    static SMRRecord deserializeFromBuffer(ByteBuf b) {
+        boolean compacted = b.readBoolean();
+        if (compacted) {
+            return COMPACTED_RECORD;
+        }
+        SMRRecord record = new SMRRecord();
+        record.serializedSize = b.readInt();
+        short methodLength = b.readShort();
+        byte[] methodBytes = new byte[methodLength];
+        b.readBytes(methodBytes, 0, methodLength);
+        record.SMRMethod = new String(methodBytes);
+        record.serializerType = Serializers.getSerializer(b.readByte());
+        byte numArguments = b.readByte();
+        Object[] arguments = new Object[numArguments];
+        for (byte arg = 0; arg < numArguments; arg++) {
+            int len = b.readInt();
+            ByteBuf objBuf = b.slice(b.readerIndex(), len);
+            arguments[arg] = record.serializerType.deserialize(objBuf, null);
+            b.skipBytes(len);
+        }
+        record.SMRArguments = arguments;
+        return record;
+    }
+
+    /**
+     * Serialize this record into a buffer.
+     * The global address is not serialized.
+     *
+     * @param b byte buffer to serialize to.
+     */
+    void serialize(ByteBuf b) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651450)